packages/core/src/render3/instructions/shared.ts

```diff
@@ -1283,15 +1286,34 @@ function findDirectiveDefMatches(
                     `Please use a different tag to activate the ${stringify(def.type)} component.`);
 
             if (isComponentHost(tNode)) {
-              // If another component has been matched previously, it's the first element in the
-              // `matches` array, see how we store components/directives in `matches` below.
-              throwMultipleComponentError(tNode, matches[0].type, def.type);
+              throwMultipleComponentError(tNode, matches.find(isComponentDef)!.type, def.type);
             }
           }
-          markAsComponentHost(tView, tNode, 0);
-          // The component is always stored first with directives after.
-          matches.unshift(def);
+
+          // Components are inserted at the front of the matches array so that their lifecycle
+          // hooks run before any directive lifecycle hooks. This appears to be for ViewEngine
+          // compatibility. This logic doesn't make sense with host directives, because it
+          // would allow the host directives to undo any overrides the host may have made.
+          // To handle this case, the host directives of components are inserted at the beginning
+          // of the array, followed by the component. As such, the insertion order is as follows:
+          // 1. Host directives belonging to the selector-matched component.
+          // 2. Selector-matched component.
+          // 3. Host directives belonging to selector-matched directives.
+          // 4. Selector-matched directives.
+          if (def.findHostDirectiveDefs !== null) {
+            const hostDirectiveMatches: DirectiveDef<unknown>[] = [];
+            // Append any host directives to the matches first.
+            def.findHostDirectiveDefs(hostDirectiveMatches, def, tView, lView, tNode);
+            hostDirectiveMatches.push(def);
+            matches.unshift(...hostDirectiveMatches);
+            markAsComponentHost(tView, tNode, hostDirectiveMatches.length - 1);
```

Comment:
nit, for readability (add `def` explicitly to the `matches` array, call out the `componentOffset` explicitly as well):
```suggestion
            // Add all host directives declared on this component, followed by the component itself.
            matches.unshift(...hostDirectiveMatches, def);
            // Component offset starting from the beginning of the host directives array.
            const componentOffset = hostDirectiveMatches.length;
            markAsComponentHost(tView, tNode, componentOffset);
```